### ISSUE TITLE
Generate Confusion matrix from a list of ground truth labels and predicted labels in yolo format

### ISSUE THREAD

**Issue by shubh-acad** [ultralytics/ultralytics#14044]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am new to yolov8 object detection and I am trying to generate confusion matrix through already generated predictions. I have ground truth labels and predicted labels in .txt format. Now I am unable to figure out how to ingest it into Yolo and run validations through val.py. Kindly guide as to how can we generate **Confusion Matrix** if we have **validation images**, **ground truth labels (.txt in yolo format)** and predicted labels( .txt in yolo format)** .

**Comment by glenn-jocher** [comment]
@shubh-acad hello,

Thank you for reaching out and welcome to the YOLOv8 community! Generating a confusion matrix from existing ground truth and predicted labels is a great way to evaluate your model's performance. Here's a step-by-step guide to help you achieve this:

1. **Prepare Your Data**: Ensure your ground truth and predicted labels are in the correct YOLO format. Each line in your `.txt` files should follow the format: `class x_center y_center width height`.

2. **Load the Labels**: You can use Python to read these `.txt` files and convert them into a format suitable for generating a confusion matrix. Below is an example of how you might do this:

    ```python
    import numpy as np
    from sklearn.metrics import confusion_matrix
    import matplotlib.pyplot as plt
    import seaborn as sns

    def load_labels(file_path):
        labels = []
        with open(file_path, 'r') as file:
            for line in file:
                labels.append(int(line.split()[0]))  # Assuming the first value is the class label
        return labels

    # Load ground truth and predicted labels
    ground_truth_labels = load_labels('path/to/ground_truth.txt')
    predicted_labels = load_labels('path/to/predicted_labels.txt')

    # Generate confusion matrix
    cm = confusion_matrix(ground_truth_labels, predicted_labels)

    # Plot confusion matrix
    plt.figure(figsize=(10, 8))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues')
    plt.xlabel('Predicted')
    plt.ylabel('Ground Truth')
    plt.title('Confusion Matrix')
    plt.show()
    ```

3. **Integrate with YOLOv8**: If you want to use the built-in functionalities of YOLOv8 for validation, you can modify the `val.py` script to read your labels and generate the confusion matrix. However, this requires a bit more customization. You can refer to the `SegmentationValidator` class in the [val.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/segment/val.py) file for guidance on how metrics are computed and plotted.

4. **Using YOLOv8's Validation Tools**: If you prefer not to modify the core YOLOv8 code, you can still use the provided tools to validate your model. Ensure your dataset is structured correctly and use the `val` mode to validate your model. YOLOv8 will automatically generate various metrics, including the confusion matrix, if configured.

    ```bash
    yolo val model=path/to/your_model.pt data=path/to/your_data.yaml
    ```

For more detailed information on validation and metrics, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/reference/models/yolo/segment/val/).

If you encounter any issues or need further assistance, please provide a minimum reproducible example of your code and data structure. This will help us better understand your setup and provide more accurate guidance. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your project! 😊
